### PR TITLE
Multiple contracts

### PR DESCRIPTION
**Pull request by PawelLechocki**
Modifications to Timber allowing it to work with multiple contracts.

TLDR:
With each request, Timber reads the `contractId` from the incoming request and builds a model with the appropriate collection name through the `assign-db-connection` middleware.

As a result, Timber will know where to insert new nodes and which collection to read the data from without mixing data from many contracts or accidentally overwriting it.

Other important changes:

Timber now requires a new environmental variable `CONTRACT_API_ENDPOINT` which is required to work with multiple contracts. It resolves the `contractId` to an actual address and provides information about the contract's deployment block (`getFromBlock()` from `merkle-tree/src/filter-controller.js`).

If it's not present, Timber will still work as usual, but will not be able to work with multiple contracts.

### CODE REVIEW

File: merkle-tree/src/filter-controller.js [hunk 1]
```diff
@@ -229,39 +228,51 @@ async function getFromBlock(db, contractName) {
...
+        `No filtering history found in mongodb, so starting filter from the contract's deployment block ${blockNumber}`,
+      );
+    }
 
-    if (transactionHash) {
-      receipt = await utilsWeb3.getTransactionReceipt(transactionHash);
-      logger.info(`receipt: ${receipt}`);
+  } else {
+    if (blockNumber === undefined) {
+      const contractInfo = await axios.get(process.env.CONTRACT_API_ENDPOINT + contractId);
```
Comment (MirandaWood) [inline merkle-tree/src/filter-controller.js:250]: Does `CONTRACT_API_ENDPOINT` require an external endpoint to be always active? If so, how does it work?

File: merkle-tree/src/routes/merkle-tree.routes.js [hunk 2]
```diff
@@ -21,48 +22,80 @@ const alreadyStarting = {}; // initialises as false
 async function startEventFilter(req, res, next) {
   logger.debug('src/routes/merkle-tree.routes startEventFilter()');
 
-  const { contractName, treeId, contractAddress } = req.body; // contractAddress & treeId are optional parameters. Address can instead be inferred by Timber in many cases.
+  const { contractName, treeId, contractId } = req.body; // contractAddress & treeId are optional parameters. Address can instead be inferred by Timber in many cases.
```
Comment (MirandaWood) [inline merkle-tree/src/routes/merkle-tree.routes.js:25]: For backwards compatibility, could we still have `contractAddress` as an optional input here? Then only use the external endpoint further down if we have a contractId

File: merkle-tree/src/routes/merkle-tree.routes.js [hunk 3]
```diff
@@ -21,48 +22,80 @@ const alreadyStarting = {}; // initialises as false
...
+  );
   // TODO: if possible, make this easier to read and follow. Fewer 'if' statements. Perhaps use 'switch' statements instead?
   try {
-    if (alreadyStarted[contractName] && (treeId === undefined || treeId === '')) {
-      res.data = { message: `filter already started for ${contractName}` };
-    } else if (alreadyStarted[(contractName, treeId)]) {
-      res.data = { message: `filter already started for ${contractName}.${treeId}` };
-    } else if (alreadyStarting[contractName] && (treeId === undefined || treeId === '')) {
+    if (
+      alreadyStarted[`${contractName} ${contractId}`] &&
```
Comment (MirandaWood) [inline merkle-tree/src/routes/merkle-tree.routes.js:34]: In the case of no `contractID` we would be storing the started info under `name ` (with a space). I'm not sure but this may turn out to be `alreadyStarted[MyContract undefined]` which is not ideal. Could we instead do something like a tuple e.g. `alreadyStarted[(contractName, contractId, treeId)]` or a string:
```
alreadyStarted[`${contractName}${contractId ? '_' + contractId : ''}${treeId ? '_' + treeId : ''}`]
```

File: merkle-tree/src/contract-controller.js [hunk 4]
```diff
@@ -105,37 +106,50 @@ async function getContractInstanceFromContractsFolder(db, contractName, contract
...
+    return contractInstance;
+  }
 }
 
 /**
 Gets a web3 contract instance a contract, and checks its consistency with the merkle tree's existing metadata db.
 @param {object} db - an instance of mongoose.createConnection (a 'Connection' instance in mongoose terminoligy). This contains permissions to access the merkle tree's databases.
 */
-async function instantiateContract(db, contractName, contractAddress) {
+async function instantiateContract(db, contractName, contractAddress, contractId) {
```
[the remark this answers is not in the document]
Comment (PawelLechocki) [reply merkle-tree/src/contract-controller.js:152]: No. We're only using it with the "default" case with `getContractInstanceFromBuildFolder()` as the build folder exists in the zapp.

Screenshot 2023-07-11 at 11 56 03

File: merkle-tree/src/db/mongodb/db.js [hunk 5]
```diff
@@ -14,29 +14,70 @@ Class created from within src/middleware/assign-db-connection
...
-        node: this.connection.model(
-          `${this.username}_${contractName}_${COLLECTIONS.NODE}`,
-          nodeSchema,
-        ),
-        metadata: this.connection.model(
-          `${this.username}_${contractName}_${COLLECTIONS.METADATA}`,
-          metadataSchema,
-        ),
-      };
+      if (contractId === undefined || contractId === '') {
```
[the remark this answers is not in the document]
Comment (PawelLechocki) [reply merkle-tree/src/db/mongodb/db.js:34]: I like this idea. While we're at it, would there be anything preventing us from getting rid of `this.username` which always resolves to `admin`? I don't see the use for it, but perhaps it'd be better to keep it for backwards compatibility?

Screenshot 2023-07-11 at 12 02 19

File: merkle-tree/src/filter-controller.js (hunk 1 above)
Comment (PawelLechocki) [reply merkle-tree/src/filter-controller.js:250]: I believe so, I can't think of any other way to find the contract's deployment block only with its address. It is possible to fetch it from ETH Json-Rpc but only if we had the transaction hash of the deployment tx. And that's just adding extra steps.

File: merkle-tree/src/routes/merkle-tree.routes.js (hunk 2 above)
Comment (PawelLechocki) [reply merkle-tree/src/routes/merkle-tree.routes.js:25]: So if `contractAddress` doesn't exist, it will still work with a single zapp.

If you follow the trace: `contractController.instantiateContract()` (line 75) -> `getContractInstanceFromBuildFolder()` -> if `contractId` doesn't exist (line 114), if `contractAddress` is `undefined` (line 118) , it goes to `utilsWeb3.getContractAddress(contractName)`and looks up the address from the Truffle's file

File: merkle-tree/src/routes/merkle-tree.routes.js (hunk 3 above)
Comment (PawelLechocki) [reply merkle-tree/src/routes/merkle-tree.routes.js:34]: Good idea, noted

File: merkle-tree/src/utils-web3.js [hunk 6]
```diff
@@ -114,21 +114,25 @@ async function getDeployedContractTransactionHash(contractName) {
...
-    contractInstance = new web3.eth.Contract(contractInterface.abi);
+    // instance:
+    if (!deployedAddress) {
+      contractInstance = new web3.eth.Contract(contractInterface.abi);
+    } else {
+      contractInstance = new web3.eth.Contract(contractInterface.abi, deployedAddress);
+    }
   } else {
     contractInstance = new web3.eth.Contract(contractInterface.abi, deployedAddress);
   }
```
[the remark this answers is not in the document]
Comment (PawelLechocki) [reply merkle-tree/src/utils-web3.js:138]: Noted

File: merkle-tree/src/contract-controller.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (MirandaWood) [reply merkle-tree/src/contract-controller.js:152]: Understood, makes sense!

File: merkle-tree/src/db/mongodb/db.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (MirandaWood) [reply merkle-tree/src/db/mongodb/db.js:34]: I believe we can remove it as well, and I would only be concerned about backwards compatibility. Since in the zapp we use mongo as a local db, setting up as admin for the user made sense, so the `.username` isn't needed.
However I'm not sure how it would work and whether it matters in a cloud deployment?

File: merkle-tree/src/filter-controller.js (hunk 1 above)
Comment (MirandaWood) [reply merkle-tree/src/filter-controller.js:250]: I see, I think calling the endpoint is fine for the starlight branch, but in main we want to keep Timber as a library which doesn't depend on anything external apart from web3. I agree that digging through the chain history to find the deployment block is just extra steps, but in future it might be good to edit `getDeployedContractTransactionHash` to take in the address and find the block some other way.

File: merkle-tree/src/routes/merkle-tree.routes.js (hunk 2 above)
Comment (MirandaWood) [reply merkle-tree/src/routes/merkle-tree.routes.js:25]: I see it would work fine - my comment was more for preventing the extra `utilsWeb3.getContractAddress(contractName)` call when the user can just pass it in if they want to. It doesn't make a huge difference but (imo) better to reduce the stack where we can.

File: merkle-tree/src/routes/merkle-tree.routes.js (hunk 3 above)
Comment (PawelLechocki) [reply merkle-tree/src/routes/merkle-tree.routes.js]: Changed to the suggested code.

File: merkle-tree/src/db/mongodb/db.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (fabiomereu-ey) [reply merkle-tree/src/db/mongodb/db.js:34]: IMHO we should keep it for now.

File: merkle-tree/src/routes/merkle-tree.routes.js (hunk 2 above)
Comment (fabiomereu-ey) [reply merkle-tree/src/routes/merkle-tree.routes.js:25]: I agree with Miranda here.

File: merkle-tree/src/filter-controller.js (hunk 1 above)
Comment (PawelLechocki) [reply merkle-tree/src/filter-controller.js:250]: We've decided to do all the fetching on the zapp's side and send the address to Timber (so Timber doesn't need to resolve it). (https://github.com/EYBlockchain/saas-transactional-ocm-zapps/tree/resolve-address)

The only issue to figure out is to how access the contract's deployment block from Timber without making any external calls.

Comment (PawelLechocki) [reply merkle-tree/src/filter-controller.js:250]: Update: We're working on fetching the block number on the zapp's side as well and passing that to Timber. That'll let us remove the last API call from the file mentioned above.

Comment (PawelLechocki) [reply merkle-tree/src/filter-controller.js]: Addressed in [0c41d29](https://github.com/EYBlockchain/timber/pull/133/commits/0c41d299b759f127662406caf1ff40c9d3c0089f)

**Review by MirandaWood — APPROVED** [review]
